### PR TITLE
enhancement: Beautify reports of conflicting `Resource` usage

### LINKED ISSUE #4826: Enhance error reporting of conflicting usages of `Resource`
With #4031 we've introduced concept of `Resource` to sinks which enables us to detect resource conflicts, like ports. We can add proper error messages which list for every resource which of the components are in conflict.

### PR DESCRIPTION
**Pull request by ktff**
Closes #4826 

Regular vector run:
```
Nov 16 00:16:38.850 ERROR vector::cli: Configuration error. error=Resource `systemd 2th socket` is claimed by multiple components: {"in2", "in4"}
Nov 16 00:16:38.850 ERROR vector::cli: Configuration error. error=Resource `127.0.0.1:8126` is claimed by multiple components: {"out2", "out3", "in3"}
Nov 16 00:16:38.850 ERROR vector::cli: Configuration error. error=Resource `stdin` is claimed by multiple components: {"in6", "in7"}
```

`validate`
```
x Resource `127.0.0.1:8126` is claimed by multiple components: {"out2", "out3", "in3"}
x Resource `stdin` is claimed by multiple components: {"in6", "in7"}
x Resource `systemd 2th socket` is claimed by multiple components: {"in2", "in4"}
```

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice. So beautiful 💅 .

File: src/sources/util/tcp.rs [hunk 1]
```diff
@@ -263,9 +263,11 @@ where
...
     match s {
         "systemd" => Ok(0),
-        s if s.starts_with("systemd#") => {
-            Ok(s[8..].parse::<usize>().map_err(de::Error::custom)? - 1)
-        }
+        s if s.starts_with("systemd#") => s[8..]
+            .parse::<usize>()
+            .map_err(de::Error::custom)?
+            .checked_sub(1)
+            .ok_or_else(|| de::Error::custom("systemd indices start from 1, found 0")),
```
Comment (bruceg) [inline src/sources/util/tcp.rs:270]: This could maybe go into `parse_system_fd` in `src/sources/util/tcp.rs`

**Review by bruceg — APPROVED** [review]
LGTM

**Review by lukesteensen — APPROVED** [review]

File: src/sources/util/tcp.rs (hunk 1 above)
Comment (ktff) [reply src/sources/util/tcp.rs:270]: Am I missing something, isn't it already there?

Comment (bruceg) [reply src/sources/util/tcp.rs:270]: My bad, I'm not sure what I was looking. I thought I had found a second place that had the dumber check for the number.

**Pull request merged**